larpandoracontent/LArTrackShowerId/TrackShowerIdFeatureTool.cc

```diff
@@ -326,19 +373,781 @@ StatusCode MipEnergyFeatureTool::ReadSettings(const TiXmlHandle xmlHandle)
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+OpeningAngleFeatureTool::OpeningAngleFeatureTool() 
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void OpeningAngleFeatureTool::Run(SupportVectorMachine::DoubleVector &featureVector, const Algorithm *const pAlgorithm,
+    const pandora::ParticleFlowObject *const pInputPfo)
+{
+    if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
+        std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
+
+    // Need the 3D clusters and hits to calculate PCA components
+    ClusterList threeDClusterList;
+    LArPfoHelper::GetThreeDClusterList(pInputPfo, threeDClusterList);
+
+    CaloHitList threeDCaloHitList; // Might be able to get through LArPfoHelper instead
+    const Cluster *const pThreeDCluster(threeDClusterList.front());
+    pThreeDCluster->GetOrderedCaloHitList().FillCaloHitList(threeDCaloHitList);
+
+    if (threeDCaloHitList.empty())
+        return;
+
+    CartesianPointVector pointVectorStart, pointVectorEnd;
+
+    this->Divide3DCaloHitList(pAlgorithm, threeDCaloHitList, pointVectorStart, pointVectorEnd);
+    // Run the PCA analysis twice
+    CartesianVector centroidStart(0.f, 0.f, 0.f), centroidEnd(0.f, 0.f, 0.f);
+    LArPcaHelper::EigenVectors eigenVecsStart, eigenVecsEnd;
+    LArPcaHelper::EigenValues eigenValuesStart(0.f, 0.f, 0.f), eigenValuesEnd(0.f, 0.f, 0.f);
+
+    LArPcaHelper::RunPca(pointVectorStart, centroidStart, eigenValuesStart, eigenVecsStart);
+    LArPcaHelper::RunPca(pointVectorEnd, centroidEnd, eigenValuesEnd, eigenVecsEnd);
+
+    float openingAngle(this->OpeningAngle(eigenVecsStart.at(0), eigenVecsStart.at(1), eigenValuesStart));
+    float closingAngle(this->OpeningAngle(eigenVecsEnd.at(0), eigenVecsEnd.at(1), eigenValuesEnd));
+
+    featureVector.push_back(std::abs(openingAngle-closingAngle));
+
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void OpeningAngleFeatureTool::Divide3DCaloHitList(const Algorithm *const pAlgorithm, CaloHitList threeDCaloHitList, CartesianPointVector &pointVectorStart, CartesianPointVector &pointVectorEnd)
+{
+    const VertexList *pVertexList = nullptr;
+    (void) PandoraContentApi::GetCurrentList(*pAlgorithm, pVertexList);
+
+    if (!pVertexList || (pVertexList->size() != 1) || (VERTEX_3D != pVertexList->front()->GetVertexType()))
+        throw StatusCodeException(STATUS_CODE_FAILURE);
+
+    const Vertex *const pVertex(pVertexList->front());
+    const CartesianVector nuVertex(pVertex->GetPosition().GetX(),pVertex->GetPosition().GetY(),pVertex->GetPosition().GetZ());
```

Comment:
Could just have ```const CartesianVector &nuVertex(pVertexList->front()->GetPosition());```

Reply:
It's funny how I overcomplicate things sometimes XD